### ISSUE TITLE
Calculating AUC (ROC and PR) with LOOCV

### ISSUE THREAD

**Issue by BU-1** [topepo/caret#822]
When using train(), how is AUC calculated for different tuning parameters when using LOOCV? Are the ROC/PR curves generated from the single sample in the validation set or are all of the samples combined to make a single curve?

**Comment by topepo** [comment]
If you have _n_ training set points, you get _n_ hold-out predictions (for each tuning parameter). The AUC is calculated from these _n_ points separately for each tuning parameter.

**Comment by BU-1** [comment]
@topepo Thanks for the response! Also wanted to say thanks for all of the great resources you've put together on the caret package. I've really learned a lot from your site and your book, Applied Predictive Modeling.

Is there a way to calculate AUC in the same manner for repeated cross-validation? I have a relatively small dataset (<1000 rows) and I would like to generate a single curve on all of the hold-out predictions, rather than generating a curve for each fold/repeat and selecting the model based on the average AUC. Do you think that would be a valid method for within-model selection on small datasets?

**Comment by topepo** [comment]
If you are using a resampling method that only holds out a training set sample once (e.g. `method = "cv"`) then you can use the `pred` object generated by `savePredictions = "final"`. 

For other resampling methods, you will have multiple holdouts per row in the training set. What I generally do is to average their predictions (the value in regression or the class probabilities in classification) so that you have one value per row. The `pred` object will contain a column for the original row number. 

Note that the ROC curve is _approximate_. You will not have exactly the same AUC as the average AUC that (non-LOO) resampling produces.

**Comment by BU-1** [comment]
I have used the `pred` object to view the resampled ROC curve after the model has been trained, but I'm interested in using the results to select the best tuning parameter. For example, if I was tuning a model using 10-fold 10-repeat cross-validation, `train` would generate predictions for 100 hold-out sets then take the average of their predictions to select the best tuning parameter. What I would like to do is combine the 100 hold-out sets together and build a single ROC curve at each tuning level, then compare these curves to select the optimal model. This is how `method = "LOOCV"` performs, but I want to use it with `repeatedcv` because `LOOCV` is computationally expensive. 

If this functionality does not exist, I'd be happy to try and add it to my code locally, but I'm not sure which functions inside the `caret` package handle this process.

Thanks again!

**Comment by topepo** [comment]
I see what you are getting at. While I don't think it is _really bad_, I don't think that it is a good idea. The way that you end up averaging the values might artificially reduce the variation in the results and give you falsely optimistic ROC values. 

The way things are set up basically enumerates the modeling process inside of resampling. It is good methodology to keep that configuration when evaluating models.

**Comment by BU-1** [comment]
Ok, thanks for the advice. I thought that if I used a really large number of folds, the variation within the resampled models would be relatively small (just like LOOCV) and I could test the performance of their results together. I understand that this process will have a high variance (just like LOOCV), but I think it could save me a lot of time while I explore different modeling types/complexities. For example, if I have 1000 samples in my training set, LOOCV would generate 1000 models at each tuning level, but if I use 50-fold 10-repeat, only 500 models will be generated and I can run my code twice as fast. 

My current plan is to set `savePredictions = "all"` and `method = "repeatedcv"` then post-process the `pred` objects with a custom function to select the optimal tuning parameter.
